ui/statistics_button.js

```diff
@@ -0,0 +1,238 @@
+      'width': 'px', 'height': 'px', 'completionPercent': 'percent',
+      'bufferingTime': 'seconds', 'drmTimeSeconds': 'seconds',
+      'licenseTime': 'seconds', 'liveLatency': 'seconds',
+      'loadLatency': 'seconds', 'manifestTimeSeconds': 'seconds',
+      'estimatedBandwidth': 'bits', 'streamBandwidth': 'bits',
+      'maxSegmentDuration': 'time', 'pauseTime': 'time', 'playTime': 'time',
+      'corruptedFrames': 'frames', 'decodedFrames': 'frames',
+      'droppedFrames': 'frames',
+    };
+
+    /** @private {!Object.<string, Function>} */
+    this.parseTo_ = {
+      'px': (name) => {
+        return this.currentStats_[name] + ' (px)';
+      },
+      'percent': (name) => {
+        return this.currentStats_[name] + ' (%)';
+      },
+      'frames': (name) => {
+        return this.currentStats_[name] + ' (frames)';
+      },
+      'seconds': (name) => {
+        return this.currentStats_[name].toFixed(2) + ' (s)';
+      },
+      'bits': (name) => {
+        return Math.round(this.currentStats_[name] / 1000) + ' (kbits/s)';
+      },
+      'time': (name) => {
+        return shaka.ui.Utils.buildTimeString(
+            this.currentStats_[name], false) + ' (m)';
+      },
+    };
+
+    /** @private {shaka.util.Timer} */
+    this.timer_ = new shaka.util.Timer(() => {
+      this.onTimerTick_();
+    });
+
+    this.updateLocalizedStrings_();
+
+    this.loadContainer_();
+
+    this.eventManager.listen(
+        this.localization, shaka.ui.Localization.LOCALE_UPDATED, () => {
+          this.updateLocalizedStrings_();
+        });
+
+    this.eventManager.listen(
+        this.localization, shaka.ui.Localization.LOCALE_CHANGED, () => {
+          this.updateLocalizedStrings_();
+        });
+
+    this.eventManager.listen(this.button_, 'click', () => {
+      this.onClick_();
+      this.updateLocalizedStrings_();
+    });
+  }
+
+  /** @private */
+  onClick_() {
+    shaka.ui.Utils.setDisplay(this.parent, false);
+
+    if (this.container_.classList.contains('shaka-hidden')) {
+      this.icon_.textContent =
+          shaka.ui.Enums.MaterialDesignIcons.STATISTICS_OFF;
+      this.timer_.tickEvery(0.1);
+      shaka.ui.Utils.setDisplay(this.container_, true);
+    } else {
+      this.icon_.textContent =
+          shaka.ui.Enums.MaterialDesignIcons.STATISTICS_ON;
+      this.timer_.stop();
+      shaka.ui.Utils.setDisplay(this.container_, false);
+    }
+  }
+
+  /** @private */
+  updateLocalizedStrings_() {
+    const LocIds = shaka.ui.Locales.Ids;
+
+    this.nameSpan_.textContent =
+        this.localization.resolve(LocIds.STATISTICS);
+
+    this.button_.ariaLabel = this.localization.resolve(LocIds.STATISTICS);
+
+    const labelText = this.container_.classList.contains('shaka-hidden') ?
+        LocIds.OFF : LocIds.ON;
+    this.stateSpan_.textContent = this.localization.resolve(labelText);
+  }
+
+  /** @private */
+  parsedStatisticValue_(name) {
+    return this.parseTo_[this.parseFrom_[name]](name);
```

Comment:
Since the statisticsList is configurable, is it possible that `this.parseFrom_[name]` returns an empty value?

Reply:
Thank you Michelle! 

The use of `this.parseFrom_[name]` is only made from elements in `this.statisticsList_`, which is previously checked to contain only valid statistics elements in `this.loadContainer_()`

This avoids calling `getConfig().statisticsList` multiple times, and also throwing errors when invalid statistics are present in the configuration, which is later tested on the `statisticsList` initialization of the tests, that contains `fakeStatistic` as an option.

However, I realized that it can throw errors when a "valid" statistic is required when there is no displayable option, which is the case of the arrays in `stateHistory` and `switchHistory`.

I added a explicit negation for those cases in my last commit!